esp_encrypted_img/src/esp_encrypted_img.c

```diff
@@ -118,13 +147,272 @@ static int decipher_gcm_key(const char *enc_gcm, esp_encrypted_img_t *handle)
+    if (err != 0) {
+        ESP_LOGE(TAG, "Failed to calculate ECC key: [0x%02X] (%s)", err, esp_err_to_name(err));
+        goto cleanup;
+    }
+
+    // Step 2: Convert output to scalar mod curve order
+    ret = mbedtls_mpi_read_binary(ecc_priv_key, hmac_output, HMAC_OUTPUT_SIZE);
+    if (ret != 0) {
+        ESP_LOGE(TAG, "failed\n  ! mbedtls_mpi_read_binary returned -0x%04x\n", (unsigned int) - ret);
+        goto cleanup;
+    }
+
+    // Reduce mod order to ensure valid private key
+    ret = mbedtls_mpi_mod_mpi(ecc_priv_key, ecc_priv_key, &order);
+    if (ret != 0) {
+        ESP_LOGE(TAG, "failed\n  ! mbedtls_mpi_mod_mpi returned -0x%04x\n", (unsigned int) - ret);
+        goto cleanup;
+    }
+
+    ESP_LOGI(TAG, "ECC key derived successfully");
+
+cleanup:
+    mbedtls_ecp_group_free(&grp);
+    mbedtls_mpi_free(&order);
+    return ret == 0 ? ESP_OK : ESP_FAIL;
+}
+
+static int derive_ota_ecc_device_key(hmac_key_id_t hmac_key, mbedtls_mpi *ecc_priv_key)
+{
+    if (!esp_encrypted_is_hmac_key_burnt_in_efuse(hmac_key)) {
+        ESP_LOGE(TAG, "Could not find HMAC key in configured eFuse block!");
+        return ESP_ERR_ENCRYPTED_IMAGE_HMAC_KEY_NOT_FOUND;
+    }
+
+    esp_err_t err = compute_ecc_key_with_hmac(hmac_key, ecc_priv_key);
+    return err;
+}
+
+static mbedtls_ecp_point *get_server_public_point(const char *data, size_t len)
+{
+    int ret;
+    uint8_t *server_public_key = NULL;
+    mbedtls_ecp_point *server_public_point = NULL;
+    mbedtls_ecp_group grp;
+    mbedtls_ecp_group_init(&grp);
+
+    if ((ret = mbedtls_ecp_group_load(&grp, MBEDTLS_ECP_DP_SECP256R1)) != 0) {
+        ESP_LOGE(TAG, "failed\n  ! mbedtls_ecp_group_load returned -0x%04x\n", (unsigned int) - ret);
+        return NULL;
+    }
+
+    server_public_point = calloc(1, sizeof(mbedtls_ecp_point));
+    if (server_public_point == NULL) {
+        ESP_LOGE(TAG, "failed to allocate memory for server public point");
+        goto cleanup;
+    }
+    mbedtls_ecp_point_init(server_public_point);
+
+    server_public_key = calloc(1, len + 1);
+    if (server_public_key == NULL) {
+        ESP_LOGE(TAG, "failed to allocate memory for server public key");
+        mbedtls_ecp_point_free(server_public_point);
+        free(server_public_key);
+        server_public_point = NULL;
+        goto cleanup;
+    }
+    server_public_key[0] = 0x04; // Uncompressed point
+    memcpy(server_public_key + 1, data, len);
+
+    ret = mbedtls_ecp_point_read_binary(&grp, server_public_point, (const unsigned char *)server_public_key, len + 1);
+    if (ret != 0) {
+        ESP_LOGE(TAG, "failed\n  ! mbedtls_ecp_point_read_binary returned -0x%04x\n", (unsigned int) - ret);
+        mbedtls_ecp_point_free(server_public_point);
+        free(server_public_key);
+        server_public_key = NULL;
+        return NULL;
+    }
+
+cleanup:
+    mbedtls_ecp_group_free(&grp);
+    if (server_public_key) {
+        memset(server_public_key, 0, len + 1);
```

Comment:
Is this memset necessary ? ( Since its a public key)

Reply:
Not necessary, but still a good idea to memset all the cryptographic artifacts back to 0.